students/MicahBraun/Lesson 04/Markov_Analysis.py

```diff
@@ -0,0 +1,100 @@
+# NAME: MICAH BRAUN
+# PROJECT: Markov_Analysis.py
+# PURPOSE: Read in string data from text file and perform data manipulations on data.
+# DATE: 06/11/2018
+# DESCRIPTION: Uses text from large body of text (I used The Brothers Karamazov) and selects a designated number of
+# text to accept as for analysis (300 in this case). Creates a dictionary of each word in the text and its associated
+# words that could follow it (based on the text available) and stores the values in the trigram format. The text is
+# then strung together "randomly" based on this predictive analysis and then printed out for the user to view.
+# ----------------------------------------------------------------------------------------------------------------------
+import re
+import random
+# --------------------------------------------------    DATA    --------------------------------------------------------
+potential_words_map = {}                           # map from root_words to a list of potential_words
+root_word = ()                                     # current tuple of words
+big_list = []                                      # used for final data display
+
+#   --------------------------------------------     FUNCTIONS     -----------------------------------------------------
+
+
+def process_f(filename, order=3):
+    """Reads-in file contents (from filename) with default value of 3 (trigram)"""
+
+    fp = open(filename)                             # open file
+
+    for line in fp:                                 # for each line of text in file: do something
+        for word in line.rstrip().split():          # for each word in each line: strip \r and split into tuple
+            process_word(word, order)               # for tuple, call process word()
+    fp.close()                                      # close file
+
+
+def process_word(word, order=3):
+    """Processes each tuple/word (str, int)
+    "During the first few iterations, all we do is store up the words;   <--- This quote is from Think Python's Example.
+    after that we start adding entries to the dictionary."               <--- I liked their description better than
+                                                                         <--- anything I could come up with.
+    """
+    global root_word                                       # access global tuple word collection
+    if len(root_word) < order:                             # if the length of the tuple is less than order(3), add-
+        root_word += (word,)                               # -word to root_word tuple. it needs to be at least 3 to work
+        return
+
+    try:
+        potential_words_map[root_word].append(word)        # add value of root_word to dict potential_words_map{}
+    except KeyError:
+        potential_words_map[root_word] = [word]            # if key of root_word does not already exist in dict, create
```

Comment:
You can avoid this try/except with `dict.setdefault` or `collections.defaultdict`